File: go/replication/replication.go
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

// Package replication provides byte-range inventory helpers for the Go CLI.
package replication

import (
	"bytes"
	"encoding/json"
	"fmt"

	"github.com/fxamacker/cbor/v2"

	"go.blackcatinformatics.ca/gts/model"
	"go.blackcatinformatics.ca/gts/reader"
	"go.blackcatinformatics.ca/gts/wire"
)

// FrameInventory records one frame's byte boundary and verified id status.
type FrameInventory struct {
	ItemIndex  int
	FrameIndex int
	Start      int
	End        int
	ID         []byte
	FrameType  string
	Valid      bool
}

// SegmentInventory records one segment's byte and fold metadata.
type SegmentInventory struct {
	Index       int
	ItemStart   int
	ItemEnd     int
	Start       int
	End         int
	Profile     string
	Head        []byte
	FrameCount  int
	Layout      model.StreamableInfo
	Diagnostics []model.Diagnostic
	Frames      []FrameInventory
}

// Inventory records the replication view of a GTS file.
type Inventory struct {
	Segments  []SegmentInventory
	Fatal     *model.Diagnostic
	Torn      int
	CleanEnd  int
	ItemCount int
}

// ByteRange is a half-open byte range.
type ByteRange struct {
	Start int
	End   int
}

// MissingStatus names the peer-head comparison result.
type MissingStatus string

// MissingComplete, MissingRanges, MissingUnknown, and MissingError describe
// the result of comparing a peer head against local inventory.
const (
	MissingComplete MissingStatus = "complete"
	MissingRanges   MissingStatus = "ranges"
	MissingUnknown  MissingStatus = "unknown"
	MissingError    MissingStatus = "error"
)

// MissingResult records byte ranges needed after a peer head.
type MissingResult struct {
	Status       MissingStatus
	FromHead     []byte
	Ranges       []ByteRange
	ScanRequired bool
	Detail       string
}

// HasProblems reports whether the inventory is not clean enough for byte resume.
func (inv *Inventory) HasProblems() bool {
	if inv.Fatal != nil || inv.Torn >= 0 {
		return true
	}
	for _, segment := range inv.Segments {
		if len(segment.Diagnostics) > 0 {
			return true
		}
	}
	return false
}

func (inv *Inventory) problemDetail() string {
	if inv.Fatal != nil {
		return fmt.Sprintf("%s: %s", inv.Fatal.Code, inv.Fatal.Detail)
	}
	if inv.Torn >= 0 {
		return fmt.Sprintf("torn at offset %d", inv.Torn)
	}
	for _, segment := range inv.Segments {
		if len(segment.Diagnostics) > 0 {
			d := segment.Diagnostics[0]
			return fmt.Sprintf("%s: %s", d.Code, d.Detail)
		}
	}
	return ""
}

func asText(v interface{}) (string, bool) {
	return wire.AsText(v)
}

func isHeaderItem(item interface{}) bool {
	inner := item
	if tag, ok := item.(cbor.Tag); ok {
		inner = tag.Content
	}
	m, ok := inner.(map[interface{}]interface{})
	if !ok {
		return false
	}
	_, hasGTS := wire.MapGet(m, "gts")
	_, hasType := wire.MapGet(m, "t")
	return hasGTS && !hasType
}

func itemEnd(items []struct {
	Offset int
	Item   interface{}
}, torn, dataLen, index int) int {
	if index+1 < len(items) {
		return items[index+1].Offset
	}
	if torn >= 0 {
		return torn
	}
	return dataLen
}

func headerProfile(item interface{}) string {
	header, err := wire.UnwrapHeader(item)
	if err != nil {
		return "generic"
	}
	if value, ok := wire.MapGet(header, "prof"); ok {
		if text, ok := asText(value); ok {
			return text
		}
	}
	return "generic"
}

func headerStoredID(item interface{}) []byte {
	header, err := wire.UnwrapHeader(item)
	if err != nil {
		return nil
	}
	if value, ok := wire.MapGet(header, "id"); ok {
		if id, ok := wire.AsBytes(value); ok {
			return id
		}
	}
	return nil
}

func headerComputedID(item interface{}) []byte {
	header, err := wire.UnwrapHeader(item)
	if err != nil {
		return nil
	}
	return wire.HeaderID(header)
}

func collectFrames(items []struct {
	Offset int
	Item   interface{}
}, torn, dataLen, start, end int) []FrameInventory {
	frames := make([]FrameInventory, 0, end-start-1)
	expectedPrev := headerStoredID(items[start].Item)
	if expectedPrev == nil {
		expectedPrev = headerComputedID(items[start].Item)
	}
	if expectedPrev == nil {
		expectedPrev = []byte{}
	}
	for itemIndex := start + 1; itemIndex < end; itemIndex++ {
		itemStart := items[itemIndex].Offset
		itemStop := itemEnd(items, torn, dataLen, itemIndex)
		frameIndex := itemIndex - start - 1
		frame, ok := items[itemIndex].Item.(map[interface{}]interface{})
		if !ok {
			frames = append(frames, FrameInventory{
				ItemIndex:  itemIndex,
				FrameIndex: frameIndex,
				Start:      itemStart,
				End:        itemStop,
				ID:         []byte{},
				FrameType:  "<non-map>",
				Valid:      false,
			})
			continue
		}
		computed := wire.ContentID(frame)
		var storedID []byte
		if value, ok := wire.MapGet(frame, "id"); ok {
			storedID, _ = wire.AsBytes(value)
		}
		frameID := computed
		if storedID != nil {
			frameID = storedID
		}
		var prev []byte
		if value, ok := wire.MapGet(frame, "prev"); ok {
			prev, _ = wire.AsBytes(value)
		}
		ftype := "<unknown>"
		if value, ok := wire.MapGet(frame, "t"); ok {
			if text, ok := asText(value); ok {
				ftype = text
			}
		}
		frames = append(frames, FrameInventory{
			ItemIndex:  itemIndex,
			FrameIndex: frameIndex,
			Start:      itemStart,
			End:        itemStop,
			ID:         frameID,
			FrameType:  ftype,
			Valid:      storedID != nil && bytes.Equal(storedID, computed) && bytes.Equal(prev, expectedPrev),
		})
		expectedPrev = frameID
	}
	return frames
}

// InventoryFor scans data into the replication inventory shape.
func InventoryFor(data []byte) *Inventory {
	items, torn := wire.IterItems(data)
	cleanEnd := len(data)
	if torn >= 0 {
		cleanEnd = torn
	}
	fs := reader.ReadFileSegments(data)
	if len(items) == 0 || fs.Fatal != nil {
		return &Inventory{
			Segments:  []SegmentInventory{},
			Fatal:     fs.Fatal,
			Torn:      torn,
			CleanEnd:  cleanEnd,
			ItemCount: len(items),
		}
	}
	var bounds []int
	for i, item := range items {
		if isHeaderItem(item.Item) {
			bounds = append(bounds, i)
		}
	}
	if len(bounds) == 0 || bounds[0] != 0 {
		return &Inventory{
			Segments:  []SegmentInventory{},
			Fatal:     fs.Fatal,
			Torn:      torn,
			CleanEnd:  cleanEnd,
			ItemCount: len(items),
		}
	}
	segments := make([]SegmentInventory, 0, len(bounds))
	for index, startItem := range bounds {
		endItem := len(items)
		if index+1 < len(bounds) {
			endItem = bounds[index+1]
		}
		if index >= len(fs.Segments) {
			break
		}
		graph := fs.Segments[index]
		start := items[startItem].Offset
		end := cleanEnd
		if endItem < len(items) {
			end = items[endItem].Offset
		}
		profile := headerProfile(items[startItem].Item)
		if len(graph.SegmentProfiles) > 0 {
			profile = graph.SegmentProfiles[0]
		}
		var head []byte
		if len(graph.SegmentHeads) > 0 {
			head = graph.SegmentHeads[0]
		}
		layout := model.StreamableInfo{}
		if len(graph.SegmentStreamable) > 0 {
			layout = graph.SegmentStreamable[0]
		}
		segments = append(segments, SegmentInventory{
			Index:       index,
			ItemStart:   startItem,
			ItemEnd:     endItem,
			Start:       start,
			End:         end,
			Profile:     profile,
			Head:        head,
			FrameCount:  endItem - startItem - 1,
			Layout:      layout,
			Diagnostics: graph.Diagnostics,
			Frames:      collectFrames(items, torn, len(data), startItem, endItem),
		})
	}
	return &Inventory{
		Segments:  segments,
		Fatal:     fs.Fatal,
		Torn:      torn,
		CleanEnd:  cleanEnd,
		ItemCount: len(items),
	}
}

func aggregateDigest(inv *Inventory) []byte {
	heads := []interface{}{}
	for _, segment := range inv.Segments {
		if segment.Head != nil {
			heads = append(heads, segment.Head)
		}
	}
	return wire.Blake3_256(wire.MustEncode([]interface{}{"gts-segment-heads-v1", heads}))
}

func optionalHex(bytes []byte) *string {
	if bytes == nil {
		return nil
	}
	text := wire.Hex(bytes)
	return &text
}

type jsonDiagnostic struct {
	Code       string `json:"code"`
	Detail     string `json:"detail"`
	FrameIndex *int   `json:"frame_index"`
}

func diagnosticJSON(d model.Diagnostic) jsonDiagnostic {
	return jsonDiagnostic{Code: d.Code, Detail: d.Detail, FrameIndex: d.FrameIndex}
}

func diagnosticsJSON(diagnostics []model.Diagnostic) []jsonDiagnostic {
	out := make([]jsonDiagnostic, 0, len(diagnostics))
	for _, diagnostic := range diagnostics {
		out = append(out, diagnosticJSON(diagnostic))
	}
	return out
}

type jsonLayout struct {
	Claimed bool    `json:"claimed"`
	Covered int     `json:"covered"`
	Tail    int     `json:"tail"`
	Head    *string `json:"head"`
}

func layoutJSON(layout model.StreamableInfo) jsonLayout {
	return jsonLayout{
		Claimed: layout.Claimed,
		Covered: layout.Covered,
		Tail:    layout.Tail,
		Head:    optionalHex(layout.Head),
	}
}

type jsonRange struct {
	Start  int `json:"start"`
	End    int `json:"end"`
	Length int `json:"length"`
}

func rangeJSON(r ByteRange) jsonRange {
	length := r.End - r.Start
	if length < 0 {
		length = 0
	}
	return jsonRange{Start: r.Start, End: r.End, Length: length}
}

func jsonLine(v interface{}) string {
	data, err := json.Marshal(v)
	if err != nil {
		panic(err)
	}
	return string(data) + "\n"
}

type jsonItemRange struct {
	Start int `json:"start"`
	End   int `json:"end"`
}

type jsonSegment struct {
	Index       int              `json:"index"`
	ByteRange   jsonRange        `json:"byte_range"`
	ItemRange   jsonItemRange    `json:"item_range"`
	Profile     string           `json:"profile"`
	Head        *string          `json:"head"`
	FrameCount  int              `json:"frame_count"`
	Layout      jsonLayout       `json:"layout"`
	Diagnostics []jsonDiagnostic `json:"diagnostics"`
}

// HeadsJSON returns the gts-replication-heads-v1 JSON document.
func HeadsJSON(inv *Inventory) string {
	segmentHeads := []string{}
	for _, segment := range inv.Segments {
		if segment.Head != nil {
			segmentHeads = append(segmentHeads, wire.Hex(segment.Head))
		}
	}
	var fileHead []byte
	if len(inv.Segments) > 0 {
		fileHead = inv.Segments[len(inv.Segments)-1].Head
	}
	var tornAt *int
	if inv.Torn >= 0 {
		torn := inv.Torn
		tornAt = &torn
	}
	var fatal *jsonDiagnostic
	if inv.Fatal != nil {
		doc := diagnosticJSON(*inv.Fatal)
		fatal = &doc
	}
	return jsonLine(struct {
		Schema       string   `json:"schema"`
		Clean        bool     `json:"clean"`
		SegmentHeads []string `json:"segment_heads"`
		Aggregate    struct {
			Schema   string  `json:"schema"`
			Count    int     `json:"count"`
			Digest   string  `json:"digest"`
			FileHead *string `json:"file_head"`
		} `json:"aggregate"`
		TornAt *int            `json:"torn_at"`
		Fatal  *jsonDiagnostic `json:"fatal"`
	}{
		Schema:       "gts-replication-heads-v1",
		Clean:        !inv.HasProblems(),
		SegmentHeads: segmentHeads,
		Aggregate: struct {
			Schema   string  `json:"schema"`
			Count    int     `json:"count"`
			Digest   string  `json:"digest"`
			FileHead *string `json:"file_head"`
		}{
			Schema:   "gts-segment-heads-v1",
			Count:    len(segmentHeads),
			Digest:   wire.Hex(aggregateDigest(inv)),
			FileHead: optionalHex(fileHead),
		},
		TornAt: tornAt,
		Fatal:  fatal,
	})
}

// SegmentsJSON returns the gts-replication-segments-v1 JSON document.
func SegmentsJSON(inv *Inventory) string {
	segments := make([]jsonSegment, 0, len(inv.Segments))
	for _, segment := range inv.Segments {
		segments = append(segments, jsonSegment{
			Index:     segment.Index,
			ByteRange: rangeJSON(ByteRange{Start: segment.Start, End: segment.End}),
			ItemRange: jsonItemRange{
				Start: segment.ItemStart,
				End:   segment.ItemEnd,
			},
			Profile:     segment.Profile,
			Head:        optionalHex(segment.Head),
			FrameCount:  segment.FrameCount,
			Layout:      layoutJSON(segment.Layout),
			Diagnostics: diagnosticsJSON(segment.Diagnostics),
		})
	}
	var tornAt *int
	if inv.Torn >= 0 {
		torn := inv.Torn
		tornAt = &torn
	}
	var fatal *jsonDiagnostic
	if inv.Fatal != nil {
		doc := diagnosticJSON(*inv.Fatal)
		fatal = &doc
	}
	return jsonLine(struct {
		Schema    string          `json:"schema"`
		Clean     bool            `json:"clean"`
		Segments  []jsonSegment   `json:"segments"`
		ItemCount int             `json:"item_count"`
		TornAt    *int            `json:"torn_at"`
		Fatal     *jsonDiagnostic `json:"fatal"`
	}{
		Schema:    "gts-replication-segments-v1",
		Clean:     !inv.HasProblems(),
		Segments:  segments,
		ItemCount: inv.ItemCount,
		TornAt:    tornAt,
		Fatal:     fatal,
	})
}

// Missing compares a peer head against local segment and frame ancestry.
func Missing(inv *Inventory, fromHead []byte) MissingResult {
	if inv.HasProblems() {
		return MissingResult{
			Status:       MissingError,
			FromHead:     fromHead,
			Ranges:       []ByteRange{},
			ScanRequired: false,
			Detail:       inv.problemDetail(),
		}
	}
	for _, segment := range inv.Segments {
		if bytes.Equal(segment.Head, fromHead) {
			ranges := []ByteRange{}
			if segment.End < inv.CleanEnd {
				ranges = append(ranges, ByteRange{Start: segment.End, End: inv.CleanEnd})
			}
			status := MissingComplete
			if len(ranges) > 0 {
				status = MissingRanges
			}
			return MissingResult{Status: status, FromHead: fromHead, Ranges: ranges, ScanRequired: false}
		}
		for _, frame := range segment.Frames {
			if frame.Valid && bytes.Equal(frame.ID, fromHead) {
				ranges := []ByteRange{}
				if frame.End < inv.CleanEnd {
					ranges = append(ranges, ByteRange{Start: frame.End, End: inv.CleanEnd})
				}
				status := MissingComplete
				if len(ranges) > 0 {
					status = MissingRanges
				}
				return MissingResult{Status: status, FromHead: fromHead, Ranges: ranges, ScanRequired: false}
			}
		}
	}
	return MissingResult{
		Status:       MissingUnknown,
		FromHead:     fromHead,
		Ranges:       []ByteRange{},
		ScanRequired: true,
		Detail:       "unknown peer head; scan required",
	}
}

// MissingJSON returns the gts-replication-missing-v1 JSON document.
func MissingJSON(result MissingResult) string {
	ranges := make([]jsonRange, 0, len(result.Ranges))
	for _, r := range result.Ranges {
		ranges = append(ranges, rangeJSON(r))
	}
	var detail *string
	if result.Detail != "" {
		detail = &result.Detail
	}
	return jsonLine(struct {
		Schema       string        `json:"schema"`
		Status       MissingStatus `json:"status"`
		FromHead     string        `json:"from_head"`
		Ranges       []jsonRange   `json:"ranges"`
		ScanRequired bool          `json:"scan_required"`
		Detail       *string       `json:"detail"`
	}{
		Schema:       "gts-replication-missing-v1",
		Status:       result.Status,
		FromHead:     wire.Hex(result.FromHead),
		Ranges:       ranges,
		ScanRequired: result.ScanRequired,
		Detail:       detail,
	})
}

// ResumeAfter returns bytes after a verified frame id and before any torn tail.
func ResumeAfter(data, frameID []byte) ([]byte, error) {
	inv := InventoryFor(data)
	if inv.HasProblems() {
		detail := inv.problemDetail()
		if detail == "" {
			detail = "input is not clean"
		}
		return nil, fmt.Errorf("%s", detail)
	}
	for _, segment := range inv.Segments {
		for _, frame := range segment.Frames {
			if frame.Valid && bytes.Equal(frame.ID, frameID) {
				return data[frame.End:inv.CleanEnd], nil
			}
		}
	}
	return nil, fmt.Errorf("frame %s not found", wire.Hex(frameID))
}

File: python/src/gts/replication.py
# SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
# SPDX-License-Identifier: MIT OR Apache-2.0
"""Replication inventory helpers for the Python CLI.

These helpers report byte ranges, segment heads, and append deltas without
folding data into a new file. They are deliberately conservative: if the input
has a fatal parse error, torn append, segment diagnostic, or broken id/prev
chain, resumable ranges are withheld and the caller is told to scan instead.
"""

from __future__ import annotations

import json
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Literal, cast

import cbor2

from gts.model import Diagnostic, StreamableInfo
from gts.reader import read_segments
from gts.wire import (
    blake3_256,
    canonical,
    content_id,
    header_id,
    iter_items,
    unwrap_header,
)


@dataclass(frozen=True)
class FrameInventory:
    """Byte and chain metadata for one non-header frame."""

    item_index: int
    frame_index: int
    start: int
    end: int
    id: bytes
    frame_type: str
    valid: bool


@dataclass(frozen=True)
class SegmentInventory:
    """Inventory for one GTS segment and its frames."""

    index: int
    item_start: int
    item_end: int
    start: int
    end: int
    profile: str
    head: bytes | None
    frame_count: int
    layout: StreamableInfo
    diagnostics: list[Diagnostic]
    frames: list[FrameInventory]


@dataclass(frozen=True)
class ByteRange:
    """Half-open byte range, suitable for HTTP Range-style replication."""

    start: int
    end: int

    @property
    def length(self) -> int:
        """Number of bytes covered by the range."""
        return max(0, self.end - self.start)


MissingStatus = Literal["complete", "ranges", "unknown", "error"]


@dataclass(frozen=True)
class MissingResult:
    """Result of comparing a peer head against local segment/frame heads."""

    status: MissingStatus
    from_head: bytes
    ranges: list[ByteRange]
    scan_required: bool
    detail: str | None = None


@dataclass(frozen=True)
class Inventory:
    """A cleanly bounded view of the file suitable for replication decisions."""

    segments: list[SegmentInventory]
    fatal: Diagnostic | None
    torn: int | None
    clean_end: int
    item_count: int

    def has_problems(self) -> bool:
        """True when replication should not trust byte-range deltas."""
        return (
            self.fatal is not None
            or self.torn is not None
            or any(segment.diagnostics for segment in self.segments)
        )

    def problem_detail(self) -> str | None:
        """Return the first human-readable reason the inventory is not clean."""
        if self.fatal is not None:
            return f"{self.fatal.code}: {self.fatal.detail}"
        if self.torn is not None:
            return f"torn at offset {self.torn}"
        for segment in self.segments:
            if segment.diagnostics:
                diagnostic = segment.diagnostics[0]
                return f"{diagnostic.code}: {diagnostic.detail}"
        return None


def _as_text(value: object) -> str | None:
    return value if isinstance(value, str) else None


def _is_header_item(item: object) -> bool:
    if isinstance(item, cbor2.CBORTag):
        item = item.value
    return isinstance(item, Mapping) and "gts" in item and "t" not in item


def _item_end(
    items: list[tuple[int, object]],
    torn: int | None,
    data_len: int,
    index: int,
) -> int:
    if index + 1 < len(items):
        return items[index + 1][0]
    return torn if torn is not None else data_len


def _header_profile(item: object) -> str:
    try:
        header = unwrap_header(item)
    except ValueError:
        return "generic"
    return _as_text(header.get("prof")) or "generic"


def _header_stored_id(item: object) -> bytes | None:
    try:
        header = unwrap_header(item)
    except ValueError:
        return None
    value = header.get("id")
    return value if isinstance(value, bytes) else None


def _header_computed_id(item: object) -> bytes | None:
    try:
        return header_id(unwrap_header(item))
    except ValueError:
        return None


def _collect_frames(
    items: list[tuple[int, object]],
    torn: int | None,
    data_len: int,
    start: int,
    end: int,
) -> list[FrameInventory]:
    frames: list[FrameInventory] = []
    expected_prev = (
        _header_stored_id(items[start][1])
        or _header_computed_id(items[start][1])
        or b""
    )
    for item_index in range(start + 1, end):
        item_start, item = items[item_index]
        item_stop = _item_end(items, torn, data_len, item_index)
        frame_index = item_index - start - 1
        if not isinstance(item, Mapping):
            frames.append(
                FrameInventory(
                    item_index=item_index,
                    frame_index=frame_index,
                    start=item_start,
                    end=item_stop,
                    id=b"",
                    frame_type="<non-map>",
                    valid=False,
                )
            )
            continue

        # iter_items decodes arbitrary CBOR maps; after the Mapping check, treat
        # keys as wire field names and validate the actual values below.
        frame = cast(Mapping[str, object], item)
        computed = content_id(frame)
        stored = frame.get("id")
        stored_id = stored if isinstance(stored, bytes) else None
        frame_id = stored_id or computed
        prev = frame.get("prev")
        valid = (
            stored_id == computed and isinstance(prev, bytes) and prev == expected_prev
        )
        expected_prev = frame_id
        frames.append(
            FrameInventory(
                item_index=item_index,
                frame_index=frame_index,
                start=item_start,
                end=item_stop,
                id=frame_id,
                frame_type=_as_text(frame.get("t")) or "<unknown>",
                valid=valid,
            )
        )
    return frames


def inventory(data: bytes) -> Inventory:
    """Parse ``data`` into segment and frame byte-range inventory."""
    items, torn = iter_items(data)
    clean_end = torn if torn is not None else len(data)
    segments, _fs_torn, fatal = read_segments(data)
    if not items or fatal is not None:
        return Inventory(
            segments=[],
            fatal=fatal,
            torn=torn,
            clean_end=clean_end,
            item_count=len(items),
        )

    bounds = [index for index, (_, item) in enumerate(items) if _is_header_item(item)]
    if not bounds or bounds[0] != 0:
        return Inventory(
            segments=[],
            fatal=fatal,
            torn=torn,
            clean_end=clean_end,
            item_count=len(items),
        )

    ends = [*bounds[1:], len(items)]
    out: list[SegmentInventory] = []
    for index, (start_item, end_item) in enumerate(zip(bounds, ends, strict=True)):
        if index >= len(segments):
            break
        graph = segments[index]
        start = items[start_item][0]
        end = items[end_item][0] if end_item < len(items) else clean_end
        out.append(
            SegmentInventory(
                index=index,
                item_start=start_item,
                item_end=end_item,
                start=start,
                end=end,
                profile=graph.segment_profiles[0]
                if graph.segment_profiles
                else _header_profile(items[start_item][1]),
                head=graph.segment_heads[0] if graph.segment_heads else None,
                frame_count=max(0, end_item - start_item - 1),
                layout=graph.segment_streamable[0]
                if graph.segment_streamable
                else StreamableInfo(),
                diagnostics=list(graph.diagnostics),
                frames=_collect_frames(items, torn, len(data), start_item, end_item),
            )
        )
    return Inventory(
        segments=out,
        fatal=fatal,
        torn=torn,
        clean_end=clean_end,
        item_count=len(items),
    )


def _aggregate_digest(inv: Inventory) -> bytes:
    heads = [segment.head for segment in inv.segments if segment.head is not None]
    return blake3_256(canonical(["gts-segment-heads-v1", heads]))


def _optional_hex(value: bytes | None) -> str | None:
    return value.hex() if value is not None else None


def _diagnostic_json(diagnostic: Diagnostic) -> dict[str, object]:
    return {
        "code": diagnostic.code,
        "detail": diagnostic.detail,
        "frame_index": diagnostic.frame_index,
    }


def _layout_json(layout: StreamableInfo) -> dict[str, object]:
    return {
        "claimed": layout.claimed,
        "covered": layout.covered,
        "tail": layout.tail,
        "head": _optional_hex(layout.head),
    }


def _range_json(byte_range: ByteRange) -> dict[str, int]:
    return {
        "start": byte_range.start,
        "end": byte_range.end,
        "length": byte_range.length,
    }


def _dumps(doc: object) -> str:
    return json.dumps(doc, separators=(",", ":")) + "\n"


def heads_json(inv: Inventory) -> str:
    """Return the stable JSON representation of segment heads."""
    segment_heads = [
        segment.head.hex() for segment in inv.segments if segment.head is not None
    ]
    file_head = inv.segments[-1].head if inv.segments else None
    return _dumps(
        {
            "schema": "gts-replication-heads-v1",
            "clean": not inv.has_problems(),
            "segment_heads": segment_heads,
            "aggregate": {
                "schema": "gts-segment-heads-v1",
                "count": len(segment_heads),
                "digest": _aggregate_digest(inv).hex(),
                "file_head": _optional_hex(file_head),
            },
            "torn_at": inv.torn,
            "fatal": _diagnostic_json(inv.fatal) if inv.fatal is not None else None,
        }
    )


def segments_json(inv: Inventory) -> str:
    """Return the stable JSON representation of segment byte ranges."""
    return _dumps(
        {
            "schema": "gts-replication-segments-v1",
            "clean": not inv.has_problems(),
            "segments": [
                {
                    "index": segment.index,
                    "byte_range": _range_json(
                        ByteRange(start=segment.start, end=segment.end)
                    ),
                    "item_range": {
                        "start": segment.item_start,
                        "end": segment.item_end,
                    },
                    "profile": segment.profile,
                    "head": _optional_hex(segment.head),
                    "frame_count": segment.frame_count,
                    "layout": _layout_json(segment.layout),
                    "diagnostics": [
                        _diagnostic_json(diagnostic)
                        for diagnostic in segment.diagnostics
                    ],
                }
                for segment in inv.segments
            ],
            "item_count": inv.item_count,
            "torn_at": inv.torn,
            "fatal": _diagnostic_json(inv.fatal) if inv.fatal is not None else None,
        }
    )


def missing(inv: Inventory, from_head: bytes) -> MissingResult:
    """Return append ranges needed after ``from_head`` or request a scan."""
    if inv.has_problems():
        return MissingResult(
            status="error",
            from_head=from_head,
            ranges=[],
            scan_required=False,
            detail=inv.problem_detail(),
        )
    for segment in inv.segments:
        if segment.head == from_head:
            ranges = (
                [ByteRange(start=segment.end, end=inv.clean_end)]
                if segment.end < inv.clean_end
                else []
            )
            return MissingResult(
                status="ranges" if ranges else "complete",
                from_head=from_head,
                ranges=ranges,
                scan_required=False,
            )
        for frame in segment.frames:
            if frame.valid and frame.id == from_head:
                ranges = (
                    [ByteRange(start=frame.end, end=inv.clean_end)]
                    if frame.end < inv.clean_end
                    else []
                )
                return MissingResult(
                    status="ranges" if ranges else "complete",
                    from_head=from_head,
                    ranges=ranges,
                    scan_required=False,
                )
    return MissingResult(
        status="unknown",
        from_head=from_head,
        ranges=[],
        scan_required=True,
        detail="unknown peer head; scan required",
    )


def missing_json(result: MissingResult) -> str:
    """Return the stable JSON representation of a missing-range result."""
    return _dumps(
        {
            "schema": "gts-replication-missing-v1",
            "status": result.status,
            "from_head": result.from_head.hex(),
            "ranges": [_range_json(byte_range) for byte_range in result.ranges],
            "scan_required": result.scan_required,
            "detail": result.detail,
        }
    )


def resume_after(data: bytes, frame_id: bytes) -> bytes:
    """Return clean trailing bytes after a validated frame id."""
    inv = inventory(data)
    if inv.has_problems():
        msg = inv.problem_detail() or "input is not clean"
        raise ValueError(msg)
    for segment in inv.segments:
        for frame in segment.frames:
            if frame.valid and frame.id == frame_id:
                return data[frame.end : inv.clean_end]
    raise ValueError(f"frame {frame_id.hex()} not found")
